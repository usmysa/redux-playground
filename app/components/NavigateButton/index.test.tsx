import { renderWithProviders } from "@/__tests__";
import { fireEvent, screen, waitFor } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router";
import { NavigateButton } from ".";

describe("NavigationButton", () => {
  it("renders correctly", () => {
    const routes = [
      {
        path: "/",
        element: <NavigateButton label="Label" to="/test" />,
      },
      {
        path: "/test",
        element: <div>TEST</div>,
      },
    ];
    const router = createMemoryRouter(routes, { initialEntries: ["/"] });
    const { container } = renderWithProviders(
      <RouterProvider router={router} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("call onBeforeNavigate only once when button is clicked", async () => {
    const onBeforeNavigate = vi.fn();
    const routes = [
      {
        path: "/",
        element: (
          <NavigateButton
            label="Label"
            onBeforeNavigate={onBeforeNavigate}
            to="/test"
          />
        ),
      },
      {
        path: "/test",
        element: <div>TEST</div>,
      },
    ];
    const router = createMemoryRouter(routes, { initialEntries: ["/"] });
    renderWithProviders(<RouterProvider router={router} />);

    const button = await screen.findByRole("button", { name: "Label" });
    fireEvent.click(button);

    await waitFor(() => {
      expect(onBeforeNavigate).toHaveBeenCalledTimes(1);
      expect(router.state.location.pathname).toEqual("/test");
    });
  });

  it("do not call onBeforeNavigate when button is not clicked", async () => {
    const onBeforeNavigate = vi.fn();
    const routes = [
      {
        path: "/",
        element: <NavigateButton label="Label" to="/test" />,
      },
      {
        path: "/test",
        element: <div>TEST</div>,
      },
    ];
    const router = createMemoryRouter(routes, { initialEntries: ["/"] });
    renderWithProviders(<RouterProvider router={router} />);

    const button = await screen.findByRole("button", { name: "Label" });
    fireEvent.click(button);

    await waitFor(() => {
      expect(onBeforeNavigate).toHaveBeenCalledTimes(0);
      expect(router.state.location.pathname).toEqual("/test");
    });
  });
});
