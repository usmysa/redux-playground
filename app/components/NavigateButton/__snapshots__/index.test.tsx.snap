// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationButton > renders correctly 1`] = `
<div>
  <button
    aria-label="Label"
    type="button"
  >
    Label
  </button>
</div>
`;
